From f430caac76c7ba140c5b53041324a374df79814f Mon Sep 17 00:00:00 2001
From: Vishal Sagar <vishal.sagar@xilinx.com>
Date: Sun, 15 Mar 2020 23:07:10 -0700
Subject: [PATCH 1278/1851] v4l: xilinx: xsdirxss: Fix format height for
 interlaced and pSF mode

commit c9c690a8b93bba7a5d19e4483932ee5b10b29882 from
https://github.com/Xilinx/linux-xlnx.git

When the transport is interlaced for 1080 line streams whose picture
type is progressive, the image format dimensions are 1920x540 instead of
1920x1080.

In 3GB DL psF mode the video is similar to interlaced as even though the
content is progressive, the transport is interlaced and is sent as two
width x (height/2) buffers.

In both cases, the field id will toggle like interlaced mode. So set the
field type as V4L2_FIELD_ALTERNATE.

Signed-off-by: Vishal Sagar <vishal.sagar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-sdirxss.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/drivers/media/platform/xilinx/xilinx-sdirxss.c b/drivers/media/platform/xilinx/xilinx-sdirxss.c
index a560814158de..29e599a00760 100644
--- a/drivers/media/platform/xilinx/xilinx-sdirxss.c
+++ b/drivers/media/platform/xilinx/xilinx-sdirxss.c
@@ -1192,6 +1192,21 @@ static int xsdirx_get_stream_properties(struct xsdirxss_state *state)
 			format->field = V4L2_FIELD_NONE;
 		else
 			format->field = V4L2_FIELD_ALTERNATE;
+
+		if (format->height == 1080 && pic_type && !tscan)
+			format->field = V4L2_FIELD_ALTERNATE;
+
+		/*
+		 * In 3GB DL pSF mode the video is similar to interlaced.
+		 * So though it is a progressive video, its transport is
+		 * interlaced and is sent as two width x (height/2) buffers.
+		 */
+		if (byte1 == XST352_BYTE1_ST372_DL_3GB) {
+			if (state->ts_is_interlaced)
+				format->field = V4L2_FIELD_ALTERNATE;
+			else
+				format->field = V4L2_FIELD_NONE;
+		}
 	}
 
 	if (format->field == V4L2_FIELD_ALTERNATE)
-- 
2.31.1

